docs/dev/zksync-v2/l1-l2-interop.md

```diff
@@ -56,13 +47,26 @@ Fourthly, there needs to be a way to bypass spam attacks for important transacti
 For each transaction, besides supplying the transaction input data, the user also provides the following:
 
 - The queue type to use. It is either a `Deque`, `HeapBuffer` or `Heap`. `Deque` is used for sequential processing of the transactions, while `HeapBuffer` and `Heap` are used for fee-prioritized processing of transactions. For the testnet, only `Deque` is available. The details on using `HeapBuffer` and `Heap` will be available later with their release.
-- The operation tree to use. It is either `Full` or `Rollup`. The details about the operation trees will be available once the zkPorter is out. For now, only the `Full` option can be used.
 - The operator fee for the transaction as part of the transaction `value` (the amount of ETH passed with the L1 transaction).
 
 ## Priority mode
 
 If the operator fails to process the needed L1 transactions, the system enters the Priority mode. In this mode, everyone can become an operator by staking tokens. The exact details of the priority mode are still under development and will be described in more detail closer to the mainnet launch.
 
-## L2->L1 communication
+## L2 -> L1 communication
+
+L2 -> L1 communication, in contrast to L1 -> L2 communication, is based only on transferring of the information, and not on the execution transaction on the other side. It is a built-in feature, which is made up of two parts: sending a message from L2 and reading a message in L1. The first is implemented as a call to the L2 system smart contract. And the second is implemented on the L1 zkSync smart contract.
+
+### Sending message
+
+Each message sent from L2 and L1 contains the sender's address and the message itself. The length of the sent message can be arbitrarily large, but the longer the message, the more expensive it is to send. The operator will "show" each message in Ethereum, so that information about all messages is always publicly available. Because of this, messages sent to L1 are guaranteed to be read, regardless of the honesty of the operator.
+
+### Reading message
+
+Every sent message can be read on-chain. Moreover, it is possible to prove that message has been sent in a specific L2 block. To make such proof as cheap as possible for both the user and the validator, we store all messages for each L2 block in the Merkle Tree. Accordingly, any L1 smart contract can consume the sent message by providing proof of inclusion to some of the L2 block. A proof can be generated based only on the data that the operator sent to the `zkSync` L1 smart contract. Proof data is also available upon API request. 
```

Comment:
```suggestion
Every message sent could be read on-chain. Moreover, it is possible to prove that a message was sent in a specific L2 block. To make such proof as cheap as possible for both the user and the validator, for each L2 block we store all messages in a merkle Tree. As a result, any L1 smart contract can consume a message by providing a proof of inclusion to some of the L2 block. This proof can be generated based only on the data that the operator sent to the `zkSync` L1 smart contract. The proof data is also available upon API request.

Reply:
> any L1 smart contract can consume a message by providing a proof of inclusion to some of the L2 block.

I don't get the last part of the sentence, it looks like something is missing. please rephrase.